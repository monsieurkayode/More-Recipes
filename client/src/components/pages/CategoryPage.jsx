import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import isEmpty from 'lodash/isEmpty';
import ReactPaginate from 'react-paginate';
import classnames from 'classnames';
import PropTypes from 'proptypes';

import {
  upvoteAction,
  downvoteAction,
  fetchRecipesByCategory,
} from '../../actions';
import { HomeNavbar, Banner } from '../headers';
import { RecipeItem, Loader, SideNav } from '../main';
import Footer from '../footer';
import materializeJavascript from '../../utils/materializeJavascript';
import categories from '../../../../shared/categories';
import pascalCase from '../../utils/pascalCase';
import resetPage from '../../utils/resetPage';
import sad from '../../../assets/css/img/sad.png';
import happySmiley from '../../../assets/css/img/happy_smiley.png';

/**
 * @summary - CategoryPage class declaration
 * @class CategoryPage
 * @extends {Component}
 */
export class CategoryPage extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf IndexPage
   */
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true,
      firstLoad: true,
      limit: 9,
      changeSmiley: false
    };
    this.toggleSmiley = this.toggleSmiley.bind(this);
    this.setPagination = this.setPagination.bind(this);
    this.hasRecipes = this.hasRecipes.bind(this);
    this.handlePageClick = this.handlePageClick.bind(this);
  }

  /**
   * @method componentDidMount
   *
   * @returns {undefined}
   */
  componentDidMount() {
    const { categoryName } = this.props.match.params;
    const { limit } = this.state;
    const currentPage = localStorage.getItem('currentCategoryPage');
    this.props.fetchRecipesByCategory(currentPage, limit, categoryName)
      .then(() => this.setState({ firstLoad: false, isLoading: false }));
  }

  /**
   * @method componentDidUpdate
   *
   * @returns {undefined}
   */
  componentDidUpdate() {
    materializeJavascript();
  }

  onClick = () => {
    resetPage();
  }

  /**
   * @method setPagination
   *
   * @param {void} void
   *
   * @returns {object} pagination props
   */
  setPagination() {
    if (this.hasRecipes()) {
      const { pagination: { page, pageCount } } = this.props.recipes;
      return {
        page,
        pageCount
      };
    }
    return {
      page: 1,
      pageCount: 1
    };
  }

  /**
   * Toggle smiley from sad face to
   * happy face
   * @method toggleSmiley
   *
   * @returns {undefined}
   */
  toggleSmiley() {
    const { changeSmiley } = this.state;
    this.setState({ changeSmiley: !changeSmiley });
  }

  /**
   * Checks if there are recipes to display
   * @method hasRecipes
   *
   * @returns {boolean} boolean
   */
  hasRecipes() {
    if (isEmpty(this.props.recipes)) {
      return false;
    }
    return !isEmpty(this.props.recipes.recipes);
  }

  /**
  * Handle page selection for pagination
  * @method handlePageClick
  *
  * @param {object} data
  *
  * @returns {void}
  */
  handlePageClick({ selected }) {
    this.setState({ isLoading: true });
    const page = selected + 1;
    const { limit } = this.state;
    localStorage.setItem('currentCategoryPage', page);
    const currentPage = localStorage.getItem('currentCategoryPage');
    const { categoryName } = this.props.match.params;
    this.props.fetchRecipesByCategory(currentPage, limit, categoryName)
      .then(() => this.setState({ isLoading: false }));
  }

  /**
  * Render recipes by category
  * @method renderCategoryRecipes
  *
  * @param {number} index
  *
  * @returns {JSX} JSX
  */
  renderCategoryRecipes = (index) => {
    const recipe = this.props.recipes.recipes[index];
    return (
      <RecipeItem
        grid="l4 m6"
        key={recipe.id}
        index={recipe.id}
        recipe={recipe}
        upvote={this.props.upvoteAction}
        downvote={this.props.downvoteAction}
      />
    );
  }

  /**
  * Render category link items
  * @method renderCategoryRecipes
  *
  * @param {number} index
  *
  * @returns {JSX} JSX
  */
  renderCategoryLinks = (index) => {
    const category = categories[index];
    const { categoryName } = this.props.match.params;
    const linkClass = classnames({ active: category === categoryName });
    return (
      <li key={category}>
        <Link
          to={`/category/${category}`}
          className={`collapsible-header truncate ${linkClass}`}
          onClick={this.onClick}
        >
          {pascalCase(category)}
        </Link>
      </li>
    );
  }

  /**
   * @method renderPagination
   *
   * @param {void} void
   *
   * @returns {JSX} JSX
   */
  renderPagination = () => {
    const { page, pageCount } = this.setPagination();
    return (
      <div className="pagination-container">
        <ReactPaginate
          previousLabel={<i className="fa fa-chevron-left" />}
          nextLabel={<i className="fa fa-chevron-right" />}
          breakLabel={<a href="">...</a>}
          breakClassName={'break-me'}
          pageCount={pageCount}
          initialPage={page - 1}
          marginPagesDisplayed={2}
          pageRangeDisplayed={5}
          onPageChange={this.handlePageClick}
          disableInitialCallback
          containerClassName={'pagination'}
          subContainerClassName={'pages pagination'}
          activeClassName={'active'}
        />
      </div>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { recipes } = this.props.recipes;
    const { isLoading, firstLoad } = this.state;
    const { categoryName } = this.props.match.params;
    return (
      <div id="category-page-container">
        { firstLoad ?
          <Loader /> :
          <div>
            <HomeNavbar />
            <Banner />
            <div>
              <h4 className="center-align category-header">
                {pascalCase(categoryName)}
              </h4>
              <div className="row">
                <div className="col l2 offset-l1 m2 hide-on-small-only">
                  <span className="category-section-header">
                    <b>Categories</b>
                  </span>
                  <p className="divider" />
                  <div className="category">
                    <ul className="collapsible z-depth-0">
                      {categories
                        .map(
                          (category, index) => this.renderCategoryLinks(index)
                        )
                      }
                    </ul>
                  </div>
                </div>
                <div className="col l8 m10" >
                  <span
                    className="recipe-section-header"
                  >
                    <b>Recipes</b>
                  </span>
                  <p className="divider" />
                  <div className="row">
                    {!isLoading ?
                      Object
                        .keys(recipes)
                        .sort((a, b) => b - a)
                        .map(index => this.renderCategoryRecipes(index))
                      :
                      <Loader />
                    }
                  </div>
                  {!this.hasRecipes() &&
                  <div className="center-align">
                    <img
                      id="smiley"
                      src={this.state.changeSmiley ?
                        happySmiley : sad}
                      alt=""
                    />
                    <h6>
                      No recipe has been created for {pascalCase(categoryName)}
                      . Care to <Link
                        onMouseEnter={this.toggleSmiley}
                        onMouseLeave={this.toggleSmiley}
                        to="/recipes/new"
                      >add</Link> one?
                    </h6>
                  </div>
                  }
                  {this.hasRecipes() && this.renderPagination()}
                </div>
              </div>
            </div>
            <SideNav />
            <Footer />
          </div>
        }
      </div>
    );
  }
}

const mapStateToProps = ({ recipes }) => ({
  pagination: recipes.pagination,
  recipes
});

CategoryPage.defaultProps = {
  recipes: {}
};

CategoryPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      categoryName: PropTypes.string.isRequired
    })
  }).isRequired,
  fetchRecipesByCategory: PropTypes.func.isRequired,
  upvoteAction: PropTypes.func.isRequired,
  downvoteAction: PropTypes.func.isRequired,
  recipes: PropTypes.shape({
    recipes: PropTypes.shape({}),
    pagination: PropTypes.shape({})
  })
};

export default connect(mapStateToProps,
  {
    upvoteAction,
    downvoteAction,
    fetchRecipesByCategory
  }
)(CategoryPage);
