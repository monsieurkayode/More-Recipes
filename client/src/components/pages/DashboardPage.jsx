import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'proptypes';

import {
  logoutAction,
  fetchUserRecipes,
  fetchUserFavorites,
  deletePost,
  removeFavorite,
  selectRecipe,
  isFetching
} from '../../actions';
import { DashboardNavbar } from '../headers';
import {
  DashboardPanel,
  SideNavDashboard,
  UserRecipe,
  UserFavoriteRecipe,
  UserProfile,
  UserNotification
} from '../dashboard';
import DeleteModal from '../modals';
import { Loader } from '../main';
import materializeJavascript from '../../utils/materializeJavascript';

/**
 * @summary - DashboardPage class declaration
 * @class DashboardPage
 * @extends {Component}
 */
export class DashboardPage extends Component {
  /**
   * @method componentDidMount
   *
   * @returns {undefined}
   */
  componentDidMount() {
    const currentPage = localStorage.getItem('currentPageUserRecipes');
    const currentPageFav = localStorage.getItem('currentPageUserFavorites');
    this.props.isFetching(true, 'Dashboard');
    this.props.fetchUserRecipes(currentPage);
    this.props.fetchUserFavorites(currentPageFav);
  }

  /**
   * @method componentDidUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidUpdate() {
    materializeJavascript();
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { isLoading } = this.props;
    const deletePostDialog = 'Delete recipe';
    const removeFavDialog = 'Remove from favorite';
    const { route } = this.props.match.params;
    return (
      <div>
        {isLoading ?
          <Loader /> :
          <div>
            <DashboardNavbar {...this.props} />
            <div className="row">
              <DashboardPanel user={this.props.user} />
              {
                route === 'profile' &&
                <UserProfile selected={route} {...this.props} />
              }
              {
                route === 'recipes' &&
                <UserRecipe selected={route} {...this.props} />
              }
              {
                route === 'favorites' &&
                <UserFavoriteRecipe selected={route} {...this.props} />
              }
              {
                route === 'notifications' &&
                <UserNotification selected={route} {...this.props} />
              }
            </div>
            <DeleteModal
              selected={route}
              id={this.props.selectedRecipe}
              dialog={route === 'recipes' ? deletePostDialog : removeFavDialog}
              handleAction={route === 'recipes' ?
                this.props.deletePost :
                this.props.removeFavorite
              }
            />
            <SideNavDashboard user={this.props.user} />
          </div>}
      </div>
    );
  }
}

const mapStateToProps = ({
  signinState,
  userRecipes,
  userFavorites,
  selectedRecipe,
  isLoading
}) => ({
  user: signinState.user,
  isAuthenticated: signinState.isAuthenticated,
  userRecipes,
  userFavorites,
  selectedRecipe,
  isLoading: isLoading.dashBoardIsLoading,
  isLoadingRecipes: isLoading.userRecipesIsLoading,
  isLoadingFavorites: isLoading.userFavoritesIsLoading
});

DashboardPage.defaultProps = {
  user: {},
  selectedRecipe: 0,
  userFavorites: {},
  userRecipes: {},
};

DashboardPage.propTypes = {
  user: PropTypes.shape({}),
  fetchUserRecipes: PropTypes.func.isRequired,
  userRecipes: PropTypes.shape({}),
  userFavorites: PropTypes.shape({}),
  fetchUserFavorites: PropTypes.func.isRequired,
  selectedRecipe: PropTypes.number,
  deletePost: PropTypes.func.isRequired,
  removeFavorite: PropTypes.func.isRequired,
  isFetching: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      route: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
};

export default connect(mapStateToProps,
  {
    logoutAction,
    fetchUserRecipes,
    fetchUserFavorites,
    deletePost,
    removeFavorite,
    selectRecipe,
    isFetching
  }
)(DashboardPage);
