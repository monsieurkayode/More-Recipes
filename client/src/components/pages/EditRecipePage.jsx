import React, { Component } from 'react';
import PropTypes from 'proptypes';

import { EditRecipe, SideNav } from '../main';
import { HomeNavbar } from '../headers';

/**
 * @summary - EditRecipePage class declaration
 * @class EditRecipePage
 * @extends {Component}
 */
class EditRecipePage extends Component {
  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    $('.dropdown-button').dropdown();
    $('.button-collapse').sideNav({
      closeOnClick: true,
      draggable: false
    });
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    return (
      <div>
        <HomeNavbar />
        <div className="row center-align">
          <div className="col s12 l6 m8 offset-l3 offset-m2">
            <h5 >Edit Recipe</h5>
          </div>
        </div>
        <EditRecipe history={this.props.history} match={this.props.match} />
        <SideNav />
      </div>
    );
  }
}

EditRecipePage.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  match: PropTypes.shape({
    isExact: PropTypes.bool.isRequired,
    params: PropTypes.shape({
      recipeId: PropTypes.string.isRequired
    }).isRequired,
    path: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
  }).isRequired
};

export default EditRecipePage;
