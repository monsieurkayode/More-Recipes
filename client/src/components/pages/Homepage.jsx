import React from 'react';
import { Link } from 'react-router-dom';

/**
 * Homepage
 *
 * @function Homepage
 *
 * @returns {JSX} JSX
 */
const Homepage = () => (
  <div className="home">
    <div className="bg-home">
      <div className="bg-overlay">
        <img className="responsive-img" src="favicon.ico" alt="" />
        <h3>Welcome to awesome tastefullness!</h3>
        <h5>Creating and sharing recipes has never been much fun</h5>
        <Link to="/signup" >
          <button id="signup" className="btn btn-home">Sign up</button>
        </Link>
        <Link to="/signin" >
          <button id="signin" className="btn btn-home">Sign In</button>
        </Link>
      </div>
    </div>
  </div>
);

export default Homepage;
