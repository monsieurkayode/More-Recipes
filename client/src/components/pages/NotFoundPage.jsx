import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'proptypes';
import resetPage from '../../utils/resetPage';
import sad from '../../../assets/css/img/sad.png';

/**
 * @summary - NotFoundPage class declaration
 * @class NotFoundPage
 * @extends {Component}
 */
class NotFoundPage extends Component {
  /**
   * Handle input change
   * @method redirectBack
   *
   * @param {void} void
   *
   * @returns {void}
   */
  redirectBack = () => {
    this.props.history.goBack();
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    return (
      <div id="not-found-container" className="center-align">
        <h2 className="red-text">
          404
        </h2>
        <img
          src={sad}
          alt=""
        />
        <h3>
          We are sorry but there is nothing here
        </h3>
        <Link to="/" onClick={resetPage}>
          <h6> <span><i className="fa fa-home" /></span> Back to home</h6>
        </Link>
        <div
          id="back-link"
          className="blue-text"
          onClick={this.redirectBack}
        >
          <span><i className="fa fa-chevron-left" /></span> Back
        </div>
      </div>
    );
  }
}

NotFoundPage.propTypes = {
  history: PropTypes.shape({
    goBack: PropTypes.func.isRequired
  }).isRequired
};

export default NotFoundPage;
