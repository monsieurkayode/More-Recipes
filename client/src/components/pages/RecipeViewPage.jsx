import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'proptypes';
import isEmpty from 'lodash/isEmpty';

import {
  fetchSingleRecipe,
  upvoteAction,
  downvoteAction,
  addFavoriteAction,
  fetchSingleFavorite,
  fetchReviews,
  removeFavorite,
} from '../../actions';
import Footer from '../footer';
import {
  Ingredients,
  Instructions,
  CommentBox,
  Comments
} from '../recipeview';
import { SideNav, Loader } from '../main';
import { HomeNavbar } from '../headers';

/**
 * @summary - RecipeViewPage class declaration
 * @class RecipeViewPage
 * @extends {Component}
 */
export class RecipeViewPage extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf RecipeViewPage
   */
  constructor(props) {
    super(props);

    this.state = {
      isLoading: true
    };
  }

  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    const { recipeId } = this.props.match.params;
    this.props.fetchSingleRecipe(recipeId)
      .then(() => this.setState({
        isLoading: false
      }));
    this.props.fetchSingleFavorite(recipeId);
    this.props.fetchReviews(recipeId);
    window.scroll(0, 0);
  }

  /**
   * @method shouldComponentUpdate
   *
   * @param {object} nextProps
   *
   * @returns {void}
   */
  shouldComponentUpdate(nextProps) {
    return !isEmpty(nextProps.currentRecipe);
  }

  /**
   * @method componentDidUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidUpdate() {
    $('.dropdown-button').dropdown();
    $('.button-collapse').sideNav({
      closeOnClick: true,
      draggable: false
    });
    $('.materialboxed').materialbox();
  }

  /**
   * Render reviews
   * @method renderReviews
   *
   * @param {number} index
   *
   * @returns {void}
   */
  renderReviews = (index) => {
    const review = this.props.reviews[index];
    return (
      <Comments
        key={review.id}
        index={review.id}
        review={review}
      />
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { currentRecipe, reviews, isFavorite } = this.props;
    const favorited = isFavorite ? 'orange-text' : '';
    const { isLoading } = this.state;
    const handleAction = isFavorite ? this.props.removeFavorite :
      this.props.addFavoriteAction;
    return (
      <div>
        { isLoading ?
          <Loader /> :
          <div>
            <HomeNavbar user={this.props.user} />
            <div id="recipe-img" className="hide-on-small-only">
              <div className="card z-depth-0">
                <div className="card-image">
                  <img
                    className="responsive-img"
                    src={currentRecipe.image}
                    alt={currentRecipe.recipeName}
                  />
                  <span className="card-title">{currentRecipe.recipeName}</span>
                </div>
              </div>
            </div>
            <div className="row">
              <Ingredients ingredients={currentRecipe.ingredients} />
              <div className="col l7 m8 s12">
                <div className="row">
                  <Instructions instructions={currentRecipe.instructions} />
                </div>
                <div className="col l12 m12 s12">
                  <div className="card recipe-view">
                    <div className="card-image">
                      <img
                        className="materialboxed"
                        src={currentRecipe.image}
                        alt=""
                      />
                      <span className="card-title right">
                        <div
                          onClick={() => this.props.upvoteAction(
                            currentRecipe.id
                          )}
                          className="chip boxReaction"
                        >
                          <i className="fa fa-thumbs-up" /> {
                            currentRecipe.upvote
                          }
                        </div>
                        <div
                          onClick={() => this.props.downvoteAction(
                            currentRecipe.id
                          )}
                          className="chip boxReaction"
                        >
                          <i className="fa fa-thumbs-down " /> {
                            currentRecipe.downvote
                          }
                        </div>
                        <div
                          onClick={() => handleAction(
                            currentRecipe.id
                          )}
                          className="chip boxReaction"
                        >
                          <i className={`fa fa-heart ${favorited}`} />
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <CommentBox />
            <div id="comment-posts" className="row">
              <div className="col l7 m8 s12 offset-l4 offset-m4">
                {
                  Object
                    .keys(reviews).sort((a, b) => b - a)
                    .map(index =>
                      this.renderReviews(index))
                }
                {!isEmpty(Object.keys(reviews)) &&
                <div className="center-align">
                  <span
                    id="view-more"
                    className="chip edit"
                  >
                    View more
                  </span>
                </div>}
              </div>
            </div>
            <SideNav />
            <Footer />
          </div>}
      </div>
    );
  }
}

RecipeViewPage.defaultProps = {
  reviews: {}
};

RecipeViewPage.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      recipeId: PropTypes.string
    })
  }).isRequired,
  fetchSingleRecipe: PropTypes.func.isRequired,
  currentRecipe: PropTypes.shape({
    id: PropTypes.number,
    views: PropTypes.number,
    upvote: PropTypes.number,
    downvote: PropTypes.number,
    recipeName: PropTypes.string,
    category: PropTypes.string,
    ingredients: PropTypes.string,
    instructions: PropTypes.string,
    image: PropTypes.string
  }).isRequired,
  upvoteAction: PropTypes.func.isRequired,
  downvoteAction: PropTypes.func.isRequired,
  addFavoriteAction: PropTypes.func.isRequired,
  removeFavorite: PropTypes.func.isRequired,
  fetchSingleFavorite: PropTypes.func.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  fetchReviews: PropTypes.func.isRequired,
  reviews: PropTypes.shape({
    id: PropTypes.number,
    userId: PropTypes.number,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string,
    comment: PropTypes.string,
    User: PropTypes.shape({
      username: PropTypes.string
    })
  })
};

const mapStateToProps = ({ currentRecipe, isFavorite, reviews }) => ({
  currentRecipe,
  isFavorite,
  reviews: reviews.comments,
  pagination: reviews.pagination
});

export default connect(mapStateToProps,
  {
    fetchSingleRecipe,
    upvoteAction,
    downvoteAction,
    addFavoriteAction,
    fetchSingleFavorite,
    fetchReviews,
    removeFavorite,
  }
)(RecipeViewPage);
