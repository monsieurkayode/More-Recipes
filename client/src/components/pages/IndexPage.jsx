import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'proptypes';

import {
  fetchRecipesAction,
  logoutAction,
  fetchTopRecipes,
} from '../../actions';
import { HomeNavbar, Banner } from '../headers';
import {
  Category,
  Recipes,
  TopRecipes,
  WelcomeMessage,
  SideNav,
  Loader
} from '../main';
import Footer from '../footer';
import Homepage from './Homepage';
import materializeJavascript from '../../utils/materializeJavascript';

/**
 * @summary - IndexPage class declaration
 * @class IndexPage
 * @extends {Component}
 */
export class IndexPage extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf IndexPage
   */
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true
    };
  }

  /**
   * @method componentDidMount
   *
   * @returns {undefined}
   */
  componentDidMount() {
    if (localStorage.getItem('token') && this.props.isAuthenticated) {
      const currentPage = localStorage.getItem('currentPage');
      this.props.fetchRecipesAction(currentPage)
        .then(() => this.setState({
          isLoading: false
        }));
      this.props.fetchTopRecipes();
    }
  }

  /**
   * @method componentDidUpdate
   *
   * @returns {undefined}
   */
  componentDidUpdate() {
    materializeJavascript();
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { isLoading } = this.state;
    const { isAuthenticated } = this.props;
    return (
      <div>
        { isAuthenticated ?
          <div>
            { isLoading ?
              <Loader /> :
              <div>
                <HomeNavbar {...this.props} />
                <Banner />
                <WelcomeMessage />
                <div className="row">
                  <Category />
                  <Recipes />
                  <TopRecipes />
                </div>
                <Footer />
                <SideNav />
              </div>}
          </div>
          :
          <Homepage />
        }
      </div>
    );
  }
}

const mapStateToProps = ({
  recipes,
  signinState
}) => ({
  recipes,
  user: signinState.user,
  isAuthenticated: signinState.isAuthenticated,
});

IndexPage.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  fetchRecipesAction: PropTypes.func.isRequired,
  fetchTopRecipes: PropTypes.func.isRequired,
  logoutAction: PropTypes.func.isRequired
};

export default connect(mapStateToProps,
  {
    fetchRecipesAction,
    logoutAction,
    fetchTopRecipes,
  }
)(IndexPage);
