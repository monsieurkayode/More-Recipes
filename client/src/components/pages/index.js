/* jshint esversion: 6 */
import SigninPage from './SigninPage';
import SignupPage from './SignupPage';
import DashboardPage from './DashboardPage';
import NotFoundPage from './NotFoundPage';
import RecipeViewPage from './RecipeViewPage';
import PostRecipePage from './PostRecipePage';
import IndexPage from './IndexPage';
import EditRecipePage from './EditRecipePage';
import HomePage from './Homepage';
import CategoryPage from './CategoryPage';

export {
  SigninPage,
  SignupPage,
  DashboardPage,
  NotFoundPage,
  RecipeViewPage,
  PostRecipePage,
  IndexPage,
  EditRecipePage,
  HomePage,
  CategoryPage
};
