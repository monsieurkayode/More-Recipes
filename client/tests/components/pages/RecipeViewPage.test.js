/* eslint-disable no-undef */
import React from 'react';

import { RecipeViewPage } from '../../../src/components/pages/RecipeViewPage';
import data from '../../__mocks__/mockData';

window.scroll = () => {};

const componentDidMountSpy = jest
    .spyOn(RecipeViewPage.prototype, 'componentDidMount'),
  componentDidUpdateSpy = jest
    .spyOn(RecipeViewPage.prototype, 'componentDidUpdate'),
  shouldComponentUpdateSpy = jest
    .spyOn(RecipeViewPage.prototype, 'shouldComponentUpdate'),
  renderSpy = jest
    .spyOn(RecipeViewPage.prototype, 'render');

const setup = (user, recipeId, isFavorite) => {
  const props = {
    user,
    isFavorite,
    reviews: {
      pagination: data.fetchReviewsResponse.pagination,
      comments: {
        [recipeId]: data.fetchReviewsResponse.comments[0]
      }
    },
    currentRecipe: data.fetchSingleRecipeResponse.recipe[0],
    removeFavorite: jest.fn(),
    addFavoriteAction: jest.fn(),
    fetchSingleRecipe: jest.fn().mockImplementation(() => Promise.resolve()),
    fetchSingleFavorite: jest.fn(),
    fetchReviews: jest.fn(),
    upvoteAction: jest.fn(),
    downvoteAction: jest.fn(),
    match: {
      params: {
        recipeId
      }
    }
  };

  const wrapper = shallow(<RecipeViewPage {...props} />);

  wrapper.setState({ isLoading: false });

  return {
    wrapper,
    props
  };
};

describe('The RecipeViewPage Component', () => {
  it('should render correctly',
    (done) => {
      const { wrapper } = setup({ username: 'username' }, '1', true);
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });

  it('should have a state of isLoading to be true when component mounts',
    (done) => {
      const { wrapper } = setup({ username: 'username' }, '1', true);
      wrapper.setState({ isLoading: true });
      expect(wrapper.state().isLoading).toBeTruthy();
      done();
    });

  it('should have a state of isLoading set to false after fetching recipe',
    (done) => {
      const { wrapper } = setup({ username: 'username' }, '1', true);
      expect(wrapper.state().isLoading).toBeFalsy();
      done();
    });

  it('should render Loader component only if isLoading state is true',
    (done) => {
      const { wrapper } = setup({ username: 'username' }, '1', true);
      wrapper.setState({ isLoading: true });
      expect(wrapper.find('Loader').length).toEqual(1);
      expect(wrapper.find('HomeNavbar').length).toEqual(0);
      expect(wrapper.find('SideNav').length).toEqual(0);
      expect(wrapper.find('Footer').length).toEqual(0);
      expect(wrapper.find('CommentBox').length).toEqual(0);
      expect(wrapper.find('Ingredients').length).toEqual(0);
      expect(wrapper.find('Instructions').length).toEqual(0);
      done();
    });

  it('should call upvoteAction when the upvote button is clicked',
    (done) => {
      const { wrapper, props } = setup({ username: 'username' }, '1', true);
      wrapper.find('.chip').first().simulate('click');
      expect(props.upvoteAction.mock.calls.length).toBe(1);
      done();
    });
  it('should call downvoteAction when the downvote button is clicked',
    (done) => {
      const { wrapper, props } = setup({ username: 'username' }, '1', true);
      wrapper.find('.chip').at(1).simulate('click');
      expect(props.downvoteAction.mock.calls.length).toBe(1);
      done();
    });

  it('should call addFavoriteAction when the favorite button is clicked' +
  ' and recipe has not been favorited',
  (done) => {
    const { wrapper, props } = setup({ username: 'username' }, '1', false);
    wrapper.find('.chip').at(2).simulate('click');
    expect(props.addFavoriteAction.mock.calls.length).toBe(1);
    done();
  });

  it('should call removeFavorite action when the favorite button is clicked' +
  ' and recipe has already been favorited',
  (done) => {
    const { wrapper, props } = setup({ username: 'username' }, '1', true);
    wrapper.find('.chip').at(2).simulate('click');
    expect(props.removeFavorite.mock.calls.length).toBe(1);
    done();
  });

  it('should call componentDidMount method',
    (done) => {
      expect(componentDidMountSpy).toHaveBeenCalled();
      done();
    });

  it('should call componentDidUpdate method',
    (done) => {
      expect(componentDidUpdateSpy).toHaveBeenCalled();
      done();
    });

  it('should call shouldComponentUpdate method',
    (done) => {
      expect(shouldComponentUpdateSpy).toHaveBeenCalled();
      done();
    });

  it('should call render method',
    (done) => {
      expect(renderSpy).toHaveBeenCalled();
      done();
    });

  describe('should have', () => {
    it('an image element with class materialboxed', (done) => {
      const { wrapper } = setup({ username: 'username' }, '1', true);
      expect(wrapper.find('.materialboxed').length).toEqual(1);
      done();
    });

    it('three buttons for upvoting, downvoting and favoriting a recipe',
      (done) => {
        const { wrapper } = setup({ username: 'username' }, '1', true);
        expect(wrapper.find('.boxReaction').length).toEqual(3);
        done();
      });

    it('a favorite button with class orange-text if recipe is favorited',
      (done) => {
        const { wrapper } = setup({ username: 'username' }, '1', true);
        expect(wrapper.find('.orange-text').exists()).toBe(true);
        done();
      });
  });
});
