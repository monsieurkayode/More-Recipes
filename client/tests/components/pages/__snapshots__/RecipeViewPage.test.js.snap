// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RecipeViewPage Component should render correctly 1`] = `
<div>
  <div>
    <Connect(HomeNavbar)
      user={
        Object {
          "username": "username",
        }
      }
    />
    <div
      className="hide-on-small-only"
      id="recipe-img"
    >
      <div
        className="card z-depth-0"
      >
        <div
          className="card-image"
        >
          <img
            alt="Chicken Soup"
            className="responsive-img"
            src="spice.jpg"
          />
          <span
            className="card-title"
          >
            Chicken Soup
          </span>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <Ingredients
        ingredients="**A list of ingredients for preparing Chicken soup**
                  * 1 (3 pound) whole chicken
                  * 4 carrots, halved
                  * 4 stalks celery, halved
                  * 1 large onion, halved
                  * water to cover
                  * salt and pepper to taste
                  * 1 teaspoon chicken bouillon granules (optional)
                  * 1/2 teaspoon lemonade"
      />
      <div
        className="col l7 m8 s12"
      >
        <div
          className="row"
        >
          <Instructions
            instructions="* Put the chicken, carrots, celery and onion in a large
                  soup pot and cover with cold water. Heat and simmer, 
                  uncovered, until the chicken meat falls off of the bones
                  (skim off foam every so often).
                  * Take everything out of the pot. Strain the broth. 
                  Pick the meat off of the bones and chop the carrots, celery 
                  and onion. Season the broth with salt, pepper and chicken 
                  bouillon to taste, if desired. Return the chicken, carrots, 
                  celery and onion to the pot, stir together, and serve."
          />
        </div>
        <div
          className="col l12 m12 s12"
        >
          <div
            className="card recipe-view"
          >
            <div
              className="card-image"
            >
              <img
                alt=""
                className="materialboxed"
                src="spice.jpg"
              />
              <span
                className="card-title right"
              >
                <div
                  className="chip boxReaction"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-thumbs-up"
                  />
                   
                  0
                </div>
                <div
                  className="chip boxReaction"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-thumbs-down "
                  />
                   
                  0
                </div>
                <div
                  className="chip boxReaction"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-heart orange-text"
                  />
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <ReduxForm />
    <div
      className="row"
      id="comment-posts"
    >
      <div
        className="col l7 m8 s12 offset-l4 offset-m4"
      >
        <Comments
          review={
            Object {
              "page": 1,
              "pageCount": 1,
              "pageSize": 1,
              "totalCount": 1,
            }
          }
        />
        <Comments
          review={
            Object {
              "1": Object {
                "User": Object {
                  "username": "username",
                },
                "comment": "Love this",
                "createdAt": "2018-02-05T07:00:47.980Z",
                "id": 1,
                "updatedAt": "2018-02-03T15:58:43.318Z",
                "userId": 1,
              },
            }
          }
        />
        <div
          className="center-align"
        >
          <span
            className="chip edit"
            id="view-more"
          >
            View more
          </span>
        </div>
      </div>
    </div>
    <Connect(SideNav) />
    <Footer />
  </div>
</div>
`;
